doc/developers/cython.rst

```diff
@@ -0,0 +1,178 @@
+
+* Activating `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_ might help. E.g. for activating boundscheck use:
+
+  .. code-block:: bash
+
+         export SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES=1
+
+* Start from scratch in a notebook to understand.
+* To debug C code (e.g. a segfault), do use ``gdb`` with:
+
+  .. code-block:: bash
+
+         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py
+
+* To have access to some value in place to debug in ``cdef (nogil)`` context, use:
+
+  .. code-block:: cython
+
+         with gil:
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+    ``print(f"{test_val=}")``
+
+* scikit-learn codebase is a bit of a zero of (fused) types (re)definitions, and we need to simplified
+  and unify that across the codebase. For now, make sure you understand which concrete types are used ultimately.
+
+* It is helpful to use ``gdb`` to debug. In order to do so, one must use a Python interpreter built with debug support
+  (debug symbols and proper optimization). To create a new conda environment with a source-built interpreter:
+
+  .. code-block:: bash
+
+         git clone https://github.com/python/cpython.git
+         conda create -n debug-scikit-dev
+         conda activate debug-scikit-dev
+         cd cpython
+         mkdir debug
+         cd debug
+         ../configure --prefix=$CONDA_PREFIX --with-pydebug
+         make EXTRA_CFLAGS='-DPy_DEBUG' -j<num_cores>
+         make install
+
+* You may need to restart conda after building/installing (I had to, but not sure why).
+
+
+  * You might find this alias handy:
+    .. code-block::
+
+         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"
+
+  * which you can use with:
+
+    .. code-block::
+
+         cythonX --annotate source.pyx
+
+Tips for performances
+^^^^^^^^^^^^^^^^^^^^^
+
+
+* Understand the GIL in context for CPython (which problems it solves, what are its limitations) and get a good
+  understanding of when Cython will be mapped to C code free of interactions with CPython, when it will not, and when
+  it cannot (e.g. presence of interactions with Python objects, which include functions).
+  In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_ provides a good overview and context and pathways for removal.
+* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
+* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryview are lightweight.
+* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and we better not use it IMO,
+  even if handling fewer dimensions in some context would be preferable.
```

Comment:
About this point, in the cython guidelines you can find https://cython-guidelines.readthedocs.io/en/latest/articles/memoryviews_are_views_not_arrays.html#slicing-is-quick

Does it mean it's fast but there's still some overhead and it's not fast enough for us ?

Also you can link to the pinned issue https://github.com/scikit-learn/scikit-learn/issues/17299

Reply:
There's definitely some overhead. It's relatively lightweight compared to (say) creating a new memoryview, and it can be done without the GIL. But it isn't free.

I assume this based on genuine hard-earned experience within Pandas.

There's also the general rule that "user code should aim to be clean and understandable, while library code might want to squeeze out every little bit of performance". I assume Pandas is more like library code in this case, so my recommendation would be a bit different than to the average user.

I'll try to add a bit more detail in the guidelines... But I think the recommendation here is reasonable